File: backend/app/game_rules/classic_rps.py
from .base import GameRuleBase
from ..models import Player, GameResult, Hand


class ClassicRPS(GameRuleBase):
    """通常じゃんけんルール"""

    def judge(self, player1: Player, player2: Player) -> GameResult:
        """勝敗判定

        Args:
            player1: プレイヤー1
            player2: プレイヤー2

        Returns:
            GameResult: プレイヤー1視点での勝敗結果
        """
        if player1.hand == player2.hand:
            return "draw"

        # 勝利条件（プレイヤー1が勝つ組み合わせ）
        win_conditions = {
            ("rock", "scissors"),
            ("scissors", "paper"),
            ("paper", "rock"),
        }

        if (player1.hand, player2.hand) in win_conditions:
            return "win"
        else:
            return "lose"

    def get_rule_name(self) -> str:
        """ルール名を取得"""
        return "Classic Rock-Paper-Scissors"

    def get_description(self) -> str:
        """ルール説明を取得"""
        return "Rock beats Scissors, Scissors beats Paper, Paper beats Rock"

File: src/game/rules/classic_rps.ts
// Classic RPS 判定関数 (Issue #7)。
//
// 純粋関数として切り出すことで、シーン (ClassicRpsScene) は描画と進行に専念し、
// 「誰が勝ったか」のロジックはここに集約する。React 版の判定と挙動を揃える。

import type { Hand, RoundResult } from '../types'

/**
 * プレイヤー (p1) 視点の勝敗を返す。
 *
 * - 同じ手 → draw
 * - rock vs scissors / scissors vs paper / paper vs rock の組で p1 が勝 → win
 * - それ以外 → lose
 */
export const judgeClassicRps = (p1: Hand, p2: Hand): RoundResult => {
  if (p1 === p2) return 'draw'
  if (
    (p1 === 'rock' && p2 === 'scissors') ||
    (p1 === 'scissors' && p2 === 'paper') ||
    (p1 === 'paper' && p2 === 'rock')
  ) {
    return 'win'
  }
  return 'lose'
}
